README.md
# Personal Web Site

<https://jonathangilligan.org/>

[![Netlify Status](https://api.netlify.com/api/v1/badges/bbf72fd4-b6e0-4e81-baef-6babad221dce/deploy-status)](https://app.netlify.com/sites/jonathangilligan/deploys)
[![License: CC BY 4.0](https://img.shields.io/badge/License-CC%20BY%204.0-lightgrey.svg)](https://creativecommons.org/licenses/by/4.0/)

The site is hosted by Netlify and built with Hugo, using my [`hugo-finisterre`](https://github.com/jonathan-g/hugo-finisterre)
theme.

[![Creative Commons License](https://mirrors.creativecommons.org/presskit/buttons/88x31/svg/by.svg)](https://creativecommons.org/licenses/by/4.0/)
This work is licensed under a [Creative Commons Attribution 4.0 International License](https://creativecommons.org/licenses/by/4.0/)


archetypes/news.md
---
title: "{{ replace .TranslationBaseName "-" " " | title }}"
author: "Jonathan Gilligan"
date: {{ .Date }}
# pub_link: 
# publishdate:
# publication:
draft: true
categories:
  - news
  - news articles
# tags:
#   - default
# summary: |
#   Summary or abstract
---


bibliography/fix_jg_pubs.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun 13 18:52:00 2017

@author: Jonathan Gilligan
"""
# import pybtex as pt
import yaml
import os
import glob
import shutil
import re
import cgi
import string
import time

def fix_files(dir):
    scratch_dir = os.path.join(dir, 'scratch')
    print "Scratch dir = ", scratch_dir
    if not os.path.isdir(scratch_dir):
        os.mkdir(scratch_dir)
    for f in os.listdir(dir):        
        dest = f.lower()
	src_file = os.path.join(dir, f)
	dest_file = os.path.join(scratch_dir, dest)
	if os.path.isfile(src_file):
	  print "Moving ", src_file, " to ", dest_file
          os.rename(src_file, dest_file)
	  redest_file = os.path.join(dir, dest)
	  print "Moving ", dest_file, " to ", redest_file
          os.rename(dest_file, redest_file)
    time.sleep(0.5)
    os.rmdir(scratch_dir)

def fix_file_ref(s):
    m = re.match("^(?P<prefix>[a-z ]+:)(?P<file>[^:]+)(?P<suffix>:.*$)", s)
    if m is not None:
        matches = m.groupdict()
        res = matches['prefix'] + matches['file'].lower() + matches['suffix']
        return res
    else:
        return s

def fix_file_refs(s):
    m = re.match("^(?P<prefix> *file *= *\\{)(?P<files>[^}]*)(?P<suffix>\\}.*$)", s)
    if m is not None:
        matches = m.groupdict()
        files = [ fix_file_ref(f) for f in string.split(matches['files'], ';') ]
        res = matches['prefix'] + string.join(files,';') + matches['suffix'] + '\n'
        return res
    else:
        return s

def process_file_refs(infile, outfile):
    with open(infile) as input:
        lines = input.readlines()
    processed_lines = [ fix_file_refs(li) for li in lines ]
    with open(outfile, 'w') as output:
        output.writelines(processed_lines)

def preprocess(infile, outfile):
    original = open(infile)
    lines = original.readlines()
    original.close()
    modified = open(outfile, 'w')
    lines = [ re.sub('^( *)[Aa]uthor\\+[Aa][Nn]', '\\1author_an', li) for li in lines ]
    lines = [ fix_file_refs(li) for li in lines ]
    modified.writelines(lines)
    modified.close()

def main():
    try:
        fix_files('pdfs')
    except RuntimeError, e:
        print e
    finally:
        pass
    process_file_refs('jg_pubs.bib', 'jg_pubs.bib')

if __name__ == '__main__':
    main()


themes/lightjason/static/js/script.js
"use strict";

jQuery(function() {

    /**
     * show-all function for fade-in all not visible category items
     */
    jQuery(".category_showall").click(function() {
        jQuery(".item_category").parent().fadeIn("slow", function() {});
    });

    /**
     * hide function for all categories
     */
    jQuery(".item_category").click(function() {
        jQuery("[data-category=" + jQuery(this).attr("data-category") + "]").parent().fadeOut("slow", function() {});
    });


    /**
     * generate toc if exists
     */
    jQuery("#toc").toc({
        "selectors": "h2,h3,h4,h5,h6",
        "container": ".body"
    });

    /**
     * hide / show toc button
     */
    jQuery("#tochideshow").click(function() {
        jQuery("#toc").fadeToggle("slow", function() {
            jQuery("#tochideshow").text().match(/hide/i) ? jQuery("#tochideshow").text("Show") : jQuery("#tochideshow").text("Hide");
        });
    });


    /**
     * navigation for small windows
     */
    jQuery("#nav").click(function() {
        if (!jQuery(".sidebar").data("active")) {
            jQuery(".sidebar").css("width", "250px");
            jQuery(".sidebar").css("margin-left", "0");
            jQuery(".sidebar").data("active", true);
        } else {
            jQuery(".sidebar").css("width", "0");
            jQuery(".sidebar").css("margin-left", "0");
            jQuery(".sidebar").data("active", false);
        }
    });


    /**
     * teletyping initialization
     */
    jQuery(".teletype").each(function() {

        jQuery(this).teletype({
            text: jQuery(this).find("p.command").map(function() {
                return jQuery(this).text().trim();
            }),
            result: jQuery(this).find("p.result").map(function() {
                return jQuery(this).html();
            }),
            prefix: jQuery(this).attr("data-prefix"),
            cursor: "\u258B",
            automaticstart: false
        });

    });

    /**
     * rest teletyping element
     */
    jQuery(".teletypereset").click(function() {
        jQuery("#" + jQuery(this).attr("data-terminal")).teletype().reset().start();
    });


    /**
     * add anchor on headlines
     */
    jQuery("h2,h3,h4,h5,h6").filter("[id]").each(function () {
        jQuery(this).html( "<a href=\"#" + jQuery(this).attr("id") + "\">" + jQuery(this).html() + "</a>" );
    });


    /**
     * tooltip toggle
     */
    jQuery(".tooltiptoggle").click( function() { 
        jQuery(this).children(".tooltipcontent:first").slideToggle(); 
    });


    /**
     * swipe-left to go to the next page
     */
    jQuery("body").hammer().bind("swipeleft", function(){
        var lc = jQuery(this).data("nextpage");
        if (lc)
            window.location.href = lc;
    });

    /**
     * swipe-right to got to the previous page
     */
    jQuery("body").hammer().bind("swiperight", function(){
        var lc = jQuery(this).data("previouspage");
        if (lc)
            window.location.href = lc;
    });

});


themes/lightjason/static/js/teletype.js
/*
 * Teletype jQuery Plugin
 * @version 0.1.6
 *
 * @author Steve Whiteley & Philipp Kraus
 * @see https://teletype.rocks
 * @see https://github.com/stvwhtly/jquery-teletype-plugin
 *
 * Copyright (c) 2015-2017 Steve Whiteley & Philipp Kraus
 * Dual licensed under the MIT or GPL Version 2 licenses.
 *
 */

"use strict";
(function () {

    var pluginname = "teletype";

    // ---- jQuery initialization -------------------------------------------------------------------------------------------

    /**
     * plugin initialize
     *
     * @param options any options
     */
    jQuery.fn[pluginname] = function (options) {
        var plugin = this.data("plugin_" + pluginname);

        if (!plugin) {
            plugin = new Teletype(this, jQuery.extend({}, jQuery.fn[pluginname].defaultSettings, options || {}));
            this.data("plugin_" + pluginname, plugin);
        }

        return plugin;
    };


    /**
     * default settings
     */
    jQuery.fn[pluginname].defaultSettings = {
        // plain text commands
        text: [],
        // results of each command as HTML data (length of text and result array must be equal, for empty results, use an empty string)
        result: [],
        // starts typing automatically otherwise with calling start-function
        automaticstart: true,
        // CSS class of command-results
        classresult: "teletype-result",
        // CSS class of teletype prefix
        classprefix: "teletype-prefix",
        // CSS class of teletype cursor
        classcursor: "teletype-cursor",
        // CSS class of output container
        classoutput: "teletype-text",
        // linebreak HTML tag
        taglinebreak: "<br/>",
        // type delay
        typeDelay: 120,
        // backward delay
        backDelay: 180,
        // cursor blink speed
        blinkSpeed: 1000,
        // cursor visualization
        cursor: "|",
        // command-line prefix
        prefix: "",
        // number foe looping typing
        loop: 1,
        // humanise typing
        humanise: true,
        // cursor smooth blinking
        smoothBlink: true,
        // callback function to catch next command (parameter full teletype DOM object)
        callbackNext: null,
        // callback function to catch typing (parameter full teletype DOM object)
        callbackType: null,
        // callback function to catch backward moving (parameter full teletype DOM object)
        vallbackBackward: null,
        // callback function which is called after finished typing (parameter full teletype DOM object)
        callbackFinished: null,
        // callback function which is calld on start typing (parameter full teletype DOM object)
        callbackStart: null,
        // callback function which is called on restting DOM element (parameter full teletype DOM object)
        callbackReset: null,
        // callback function, which is called on next loop (parameter full teletype DOM object)
        callbackNextLoop: null
    };


    // ---- plugin definition -----------------------------------------------------------------------------------------------

    /**
     * plugin factory
     *
     * @param po_element DOM element
     * @param po_options initialize options
     */
    function Teletype(po_element, po_options) {
        this.dom = po_element;
        this.settings = po_options;

        return initialize(this);
    }


    Teletype.prototype = {

        /**
         * modifies the internal cursor representation
         *
         * @param pc_cursor cursor character
         * @return self reference
         */
        cursor: function (pc_cursor) {
            this.settings.cursor = pc_cursor;
            this.dom.find("." + this.settings.classcursor + ":first").html(this.settings.cursor);

            return this;
        },


        /**
         * resets the dom element with clearning
         *
         * @return self reference
         */
        reset: function () {
            if (this.settings.loop === 0) {
                return this;
            }

            setCurrentString(clearCurrent(this.stop().empty()));

            if (typeof(this.settings.callbackReset) === "function") {
                this.settings.callbackReset(this);
            }

            return this;
        },


        /**
         * clears the text-field of output
         *
         * @return self reference
         */
        empty: function () {
            this.dom.find("." + this.settings.classoutput + ":first").empty();
            return this;
        },


        /**
         * stops current animation
         *
         * @return self reference
         */
        stop: function () {
            if (this.current.timeout) {
                clearTimeout(this.current.timeout);
            }

            return this;
        },


        /**
         * starts typing if automatic start is disabled
         *
         * @return self reference
         */
        start: function () {
            if (this.settings.automaticstart) {
                return this;
            }

            setCurrentString(this);
            if (typeof(this.settings.callbackStart) === "function") {
                this.settings.callbackStart(this);
            }

            type(this);
            return this;
        }

    };


    // ---- private function ------------------------------------------------------------------------------------------------

    /**
     * sets the current string and if possible result data
     *
     * @param po_this execution context
     * @return self reference
     */
    var setCurrentString = function (po_this) {
        if ((!po_this.settings.text) || (po_this.settings.text.length === 0)) {
            return po_this;
        }

        po_this.current.string = po_this.settings.text[po_this.current.index].replace(/\n/g, "\\n");
        po_this.current.letters = po_this.current.string.split("");
        po_this.current.result = ((po_this.settings.result.length === po_this.settings.text.length) && (po_this.settings.result[po_this.current.index]))
                                    ? "<p class=\"" + po_this.settings.classresult + "\">" + po_this.settings.result[po_this.current.index] + "</p>"
                                    : "";

        return po_this;
    };


    /**
     * clear current
     *
     * @param po_this execution context
     * @return self reference
     */
    var clearCurrent = function (po_this) {
        po_this.current = {
            string: "",
            result: "",
            letters: [],
            index: 0,
            position: 0,
            loop: 0,
            timeout: null
        };

        return po_this;
    };


    /**
     * delay function
     *
     * @param po_this execution context
     * @param pn_speed any speed value
     * @return randomized speed value
     */
    var delay = function (po_this, pn_speed) {
        return po_this.settings.humanise
                ? Math.round(parseInt(pn_speed) + Math.random() * pn_speed / 3)
                : parseInt(pn_speed);
    };


    /**
     * extract a number from a text
     *
     * @param pc_text input text
     * @param pn_start start position within the string
     * @return extracted number
     */
    var extractnumber = function (pc_text, pn_start) {
        var end = pc_text.substr(pn_start).search(/[^0-9]/);
        return pc_text.substr(
                    pn_start, end === -1
                    ? pc_text.length
                    : end
               );
    };


    /**
     * pause function for typing pause
     *
     * @param po_this execution context
     * @param pn_start start position for searching pause time
     * @return self reference
     */
    var pause = function (po_this, pn_start) {
        var time = extractnumber(po_this.current.string, pn_start);
        if (!jQuery.isNumeric(time)) {
            return po_this;
        }

        po_this.current.position = pn_start + time.length;
        po_this.current.timeout = setTimeout(type(po_this), time);

        return po_this;
    };


    /**
     * sets the next outpur sequence
     *
     * @param po_this execution context
     * @return boolean next line exists
     */
    var next = function (po_this) {
        po_this.current.index += 1;

        // check end and looping
        if (po_this.current.index >= po_this.settings.text.length) {
            po_this.current.index = 0;
            po_this.current.loop += 1;

            if (typeof(po_this.settings.callbackNextLoop) === "function") {
                po_this.settings.callbackNextLoop(po_this);
            }

            if ((po_this.settings.loop !== false) && (po_this.settings.loop === po_this.current.loop)) {
                // runs finished callback
                if (typeof(po_this.settings.callbackFinished) === "function") {
                    po_this.settings.callbackFinished(po_this);
                }

                return false;
            }
        }

        setCurrentString(po_this);
        po_this.current.position = 0;

        // runs next-callback
        if (typeof(po_this.settings.callbackNext) === "function") {
            po_this.settings.callbackNext(po_this);
        }

        return true;
    };


    /**
     * constructor
     *
     * @param po_this execution context
     * @return self reference
     */
    var initialize = function (po_this) {

        // clear DOM node first
        po_this.dom.empty();

        // sets instance an nessessary DOM values into element
        clearCurrent(po_this);
        po_this.output = jQuery("<span/>").addClass(po_this.settings.classoutput).appendTo(po_this.dom);

        // set cursor
        if (po_this.settings.cursor) {
            var cursor = jQuery("<span/>").addClass(po_this.settings.classcursor).appendTo(po_this.dom).text(po_this.settings.cursor);
            var self = po_this;
            setInterval(function () {
                if (self.settings.smoothBlink) {
                    cursor.animate({
                        opacity: 0
                    }).animate({
                        opacity: 1
                    });
                } else {
                    cursor.delay(500).fadeTo(0, 0).delay(500).fadeTo(0, 1);
                }
            }, po_this.settings.blinkSpeed);
        }

        // start typing
        if (po_this.settings.automaticstart) {
            setCurrentString(po_this);
            if (typeof(po_this.settings.callbackStart) === "function") {
                po_this.settings.callbackStart(po_this);
            }

            type(po_this);
        }

        return po_this;
    };


    /**
     * backspace for removing characters
     *
     * @param po_this execution context
     * @param pn_stop number of characters to remove
     * @return self reference
     */
    var backspace = function (po_this, pn_stop) {
        if ((pn_stop < 1) || (po_this.current.position - pn_stop < 1)) {
            po_this.current.timeout = setTimeout(type(po_this), delay(po_this, po_this.settings.typeDelay));
            return po_this;
        }

        po_this.current.timeout = setTimeout(function () {
            po_this.output.html(po_this.output.html().slice(0, -1));
            backspace(po_this, pn_stop - 1);

            if (typeof(po_this.settings.callbackBackward) === "function") {
                po_this.settings.callbackBackward(po_this);
            }

        }, delay(po_this, po_this.settings.backDelay));

        return po_this;
    };


    /**
     * execution typing
     *
     * @param po_this execution context
     */
    var type = function (po_this) {

        // add new prefix item if possible
        if ((po_this.settings.prefix) && (po_this.current.position === 0)) {
            jQuery("<span />").addClass(po_this.settings.classprefix).html(po_this.settings.prefix).appendTo(po_this.output);
        }

        // get current letter & position
        var letter = po_this.current.letters[po_this.current.position];
        var start = po_this.current.position + 1;

        // check pause character
        if (letter === "^") {
            pause(po_this, start);
            return;
        }

        // check for backward character
        if (letter === "~") {
            var value = extractnumber(po_this.current.string, start);
            if (jQuery.isNumeric(value)) {
                var self = po_this;
                po_this.current.position += value.length + 1;
                po_this.current.timeout = setTimeout(function () {
                    backspace(self, value);
                }, delay(po_this, po_this.settings.backDelay * value));
                return;
            }
        }

        // check for line-break
        if ((letter === "\\") && (po_this.current.string.substr(start, 1) === "n")) {
            po_this.current.position += 1;
            letter = po_this.settings.taglinebreak;
        }


        // run typing-callback
        if (typeof(po_this.settings.callbackType) === "function") {
            po_this.settings.callbackType(po_this);
        }


        // run the next iteration
        if (po_this.current.position < po_this.current.string.length) {

            po_this.current.timeout = setTimeout(function () {

                // increment current position and set output
                po_this.current.position += 1;
                po_this.output.html(po_this.output.html() + letter);

                type(po_this);
            }, delay(po_this, po_this.settings.typeDelay));

        } else {

            // set the result (of the typing) if it exists
            if (po_this.current.result) {
                po_this.output.html(po_this.output.html() + po_this.current.result);
            }

            // check if there exists a new line
            if (next(po_this)) {
                po_this.output.html(po_this.output.html() + po_this.settings.taglinebreak);
                po_this.current.timeout = setTimeout(type(po_this), delay(po_this, po_this.settings.typeDelay));
            }

        }

    };

}(jQuery));


bibliography/process_bibliography.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun 13 18:52:00 2017

@author: Jonathan Gilligan
"""
# import pybtex as pt
import os
import glob
import time
import sys
import shutil
import subprocess
import re
import html
import yaml
import urllib.parse
# import string
import datetime
from dateutil.parser import parse as duparse
import pybtex.database as ptd

def fix_files(path):
    scratch_path = os.path.join(path, 'scratch')
    if not os.path.isdir(scratch_path):
        os.mkdir(scratch_path)
    for f in os.listdir(path):
        dest = f.lower()
        os.rename(os.path.join(path, f), os.path.join(scratch_path, dest))
        os.rename(os.path.join(scratch_path, dest), os.path.join(path, dest))
    time.sleep(0.5)
    os.rmdir(scratch_path)

fix_file_pattern_1 = re.compile("^(?P<prefix>[a-z ]+:)(?P<file>[^:]+)(?P<suffix>:.*$)")
fix_file_pattern_2 = re.compile("^(?P<prefix> *file *= *\\{)(?P<files>[^}]*)(?P<suffix>\\}.*$)")

def fix_file_ref(s):
    m = fix_file_pattern_1.match(s)
    if m is not None:
        matches = m.groupdict()
        res = matches['prefix'] + matches['file'].lower() + matches['suffix']
        return res
    else:
        return s

def fix_file_refs(s):
    m = fix_file_pattern_2.match(s)
    if m is not None:
        matches = m.groupdict()
        files = [ fix_file_ref(f) for f in matches['files'].split(';') ]
        res = matches['prefix'] + ';'.join(files) + matches['suffix'] + '\n'
        return res
    else:
        return s

def process_file_refs(infile, outfile):
    with open(infile, encoding="utf-8") as source:
        lines = source.readlines()
    processed_lines = [ fix_file_refs(li) for li in lines ]
    with open(outfile, 'w', encoding="utf-8") as sink:
        sink.writelines(processed_lines)

patent_strip_pattern = re.compile("[^A-Za-z0-9]+")

def patent_strip(s):
    s = re.sub(patent_strip_pattern, "", s).upper()
    return s

def preprocess(infile, outfile):
    original = open(infile, encoding="utf-8")
    lines = original.readlines()
    original.close()
    modified = open(outfile, 'w', encoding="utf-8")
    lines = [ re.sub('^( *)[Aa]uthor\\+[Aa][Nn]', '\\1author_an', li) for li in lines ]
    lines = [ fix_file_refs(li) for li in lines ]
    modified.writelines(lines)
    modified.close()

def call_citeproc(source, target):
    os.system('pandoc ' + source + ' -s -f biblatex -t markdown > ' + target)
    f = open(target, 'rt')
    lines = f.readlines()
    lines = [l for l in lines if l != '---\n']
    f.close()
    f = open(target, 'wt')
    f.writelines(lines)
    f.close()

file_expr = re.compile('^(?P<desc>[^:]*):(?P<path>[^:]+)[\\\\/](?P<file>[^:/\\\\]+):(?P<type>[^:;]*)$')

def extract_file_link(filestr):
    files = filestr.split(';')
    matches = [file_expr.match(s) for s in files ]
    # d = dict([(m.group('desc'), m.group('file')) for m in matches])
    for i in range(len(files)):
        if matches[i] is None:
            print('ERROR: cannot match file string "%s" from "%s".' % (files[i], filestr))
    d = [ {'desc':m.group('desc'), 'file': m.group('file')} for m in matches]
    return d

def merge(bitem, yitem):
    fields = ['file', 'title_md', 'booktitle_md', 'note_md', 'amazon', 'preprint', 'ssrn', 'patent_num']

    for f in fields:
        if f in bitem.fields.keys():
            s = str(bitem.fields[f])
            if f == 'file':
                yitem[f] = extract_file_link(s)
            else:
                yitem[f] = s
    return yitem

def process_item(bitem, yitem):
    yitem = merge(bitem, yitem)
    return yitem

def gen_refs(bibfile):
    target = os.path.splitext(os.path.split(bibfile)[1])[0] + '.yml'
    call_citeproc(bibfile, target)

    bib = ptd.parse_file(bibfile)
    ybib = yaml.safe_load(open(target, encoding = 'utf-8'))

    for yitem in ybib['references']:
        bitem = bib.entries.get(yitem['id'])
        yitem = merge(bitem, yitem)

    yaml.dump(ybib, open('publications.yml', 'w', encoding="utf-8"))
    return ybib

clean_expr = re.compile('[^a-zA-z0-9]+')

given_s1_expr = re.compile('\\b([A-Z])[a-z]+\\b')
given_s2_expr = re.compile('\\. +')

def shorten_given(s, f):
    s1 = re.sub(given_s1_expr, '\\1.', s)
    s2 = re.sub(given_s2_expr, '.', s1)
    # print('item ', key, ': fam = "', f, '", s = "', s, '", s1 = "', s1, '", s2 = "', s2, '"')
    return s2

def fix_particles(name):
    if ('family' in name.keys()):
        if ('non-dropping-particle' in name.keys()):
            # print('item ', key, ': *** fam = "', name['family'], '", nd part = "', 
            #       name['non-dropping-particle'], '"')
            name['family'] = name['non-dropping-particle'].rstrip() + ' ' +  \
                name['family'].lstrip()
            name.pop('non-dropping-particle')
        if ('dropping-particle' in name.keys()):
            # print('item ', key, ': *** fam = "', name['family'], '", d part = "', 
            #       name['dropping-particle'], '"')
            name['family'] = name['dropping-particle'].rstrip() + ' ' +  \
                name['family'].lstrip()
            name.pop('dropping-particle')
    return name

def gen_items(bib):
    output_keys = ['title', 'author', 'short_author',
                   'short_title',
                   'container_title', 'collection_title',
                   'editor', 'short_editor',
                   'publisher_place', 'publisher',
                   'genre', 'status',
                   'volume', 'issue', 'page', 'number',
                   'ISBN', 'DOI', # 'URL',
                   'pub_url',
                   'preprint',
                   'ssrn',
                   'patent_num',
                   'issued',
                   'keyword',
                   'note',
                   'file',
                   'amazon'
                   ]
    # title_keys = ['title', 'short_title', 'container_title', 'collection_title']
    proxy_pattern = re.compile(r'^(?P<hostname>[a-z0-9-]+)(?P<proxy>\.proxy(\.[a-z0-9-]+)*\.vanderbilt\.edu)')
    abstract_pattern = re.compile(r"(?P<keep>(?P<emph>([*_]+|&quot;))(?P<text>.+)(?P=emph))(?P=text)")
    if not os.path.exists('content'):
        os.mkdir('content')
    for item in bib:
        keys = item.keys()
        key = clean_expr.sub('_', item['id'])
        # print("Item ", key, " has keys ", ', '.join(keys))
        if 'url' in keys: 
          # print(" item ", key, " has url ", item['url'])
          pass
        if 'doi' in keys:
          item['DOI'] = item.pop('doi')
          if 'url' in item.keys():
            # print("Deleting URL for ", key)
            item.pop('url')
        if 'title-short' in keys:
            item['short_title'] = item['title-short']
        if 'container-title' in keys:
            item['container_title'] = item['container-title']
        if 'collection-title' in keys:
            item['collection_title'] = item['collection-title']
        if 'publisher-place' in keys:
            item['publisher_place'] = item['publisher-place']
        if 'author' in keys:
            item['author'] = [ fix_particles(n) for n in item['author']]
            item['short_author'] = [ {'family':n['family'], 'given':shorten_given(n['given'], n['family'])} for n in item['author'] ]
        if 'editor' in keys:
            item['editor'] = [ fix_particles(n) for n in item['editor']]
            item['short_editor'] = [ {'family':n['family'], 'given':shorten_given(n['given'], n['family'])} for n in item['editor'] ]
        header_items = dict([(k, v) for (k, v) in item.items() if k in output_keys])
        # for tk in title_keys:
        #     if (tk in header_items.keys()):
        #         header_items[tk] = re.sub('\\^([^\\^]+)\\^', '<sup>\\1</sup>', header_items[tk])
        header_items['id'] = key
        if ('issued' in header_items.keys()):
          issued = header_items['issued']
          if isinstance(issued, str):
            dt = duparse(issued)
            dd = { 'year': dt.year, 'month': dt.month, 'day': dt.day}
            print("dd = ", dd)
          elif isinstance(issued, tuple) or isinstance(issued, list):
            dd = header_items['issued'][0]
          elif isinstance(issued, datetime.date):
            dd = {'year': issued.year, 'month': issued.month, 'day': issued.day}
          elif isinstance(issued, int):
            dd = {'year': issued}
          else:
            print("issued: type = ", type(issued), ", value = ", issued)
            exit(1)
          header_items['issued'] = dd
          y = int(dd['year'])
          m = 1
          d = 1
          if 'month' in dd.keys():
              m = int(dd['month'])
          if 'day' in dd.keys():
            d = int(dd['day'])
          header_items['date'] = ("%04d-%02d-%02d" % (y, m, d))
          d = datetime.datetime.strptime(header_items['date'], "%Y-%m-%d").date()
          if (d > datetime.date.today()):
              d = datetime.date.today()
          header_items['pubdate'] = d.isoformat()
        else :
          print("No issued date for ", header_items['id'])
          continue
        if 'url' in keys:
            # print("Item ", key, " has an URL")
            if not 'doi' in keys:
              # print("Fixing up URL for item ", key)
              pub_url = item['url']
              split_url = urllib.parse.urlsplit(pub_url)
              hostname = split_url.netloc
              m = proxy_pattern.match(hostname)
              if m:
                gd = m.groupdict()
                if 'proxy' in gd.keys():
                  hostname = gd['hostname'].replace('-', '.')
                  split_url = split_url._replace(netloc = hostname)
                  pub_url = urllib.parse.urlunsplit(split_url)
                  # print("Replacing proxy for item ", key)
              # print("URL for ", key, " is ", pub_url)
              header_items['pub_url'] = pub_url
            else:
              # print("Ignoring URL for item ", key)
              pass
        if 'preprint' in keys:
            header_items['preprint_url'] = item['preprint']
        if 'ssrn' in keys:
            header_items['ssrn_id'] = item['ssrn']
        header_items['pub_type'] = item['type']
        if 'keyword' in header_items.keys():
            k = header_items['keyword']
            k = re.sub("[a-z\\- ]*tenure[a-z\\- ]*", "", k)
            k = re.sub(", *,", ",", k)
            k = re.sub("^, *", "", k)
            k = re.sub(" *, *$", "", k)
            k = k.strip()
            if (len(k) == 0):
                del header_items['keyword']
            else:
                header_items['keyword'] = k
        if (item['type'] == 'patent' and 'number' in item.keys()):
            header_items['patent_num'] = patent_strip(item['number'])
        outfile = open(os.path.join('content', key + '.md'), 'w', encoding="utf-8")
        outfile.write('---\n')
        yaml.dump(header_items, outfile)
        outfile.write('---\n')
        abstract = None
        if 'abstract_md' in item.keys():
            abstract = item['abstract_md']
        elif 'abstract' in item.keys():
            abstract = item['abstract']
        if abstract is not None:
            abstract = html.escape(abstract).encode('ascii', 'xmlcharrefreplace').decode('utf-8')
            abstract = abstract_pattern.sub(r'\g<keep>', abstract)
            outfile.write(abstract + '\n')
        outfile.close()

def move_md_files(src = 'content', dest = '../content/publications'):
    files = glob.glob(os.path.join(src, '*.md'))
    if not os.path.isdir(dest):
        os.makedirs(dest)
    for f in files:
        base = os.path.split(f)[1]
        dest_file = os.path.join(dest, base)
        shutil.copyfile(f, dest_file)

def move_pdf_files(src = 'pdfs', dest = '../static/files/pubs/pdfs'):
    files = os.listdir(src)
    if not os.path.isdir(dest):
        os.makedirs(dest)
    for f in files:
        src_file = os.path.join(src, f)
        dest_file = os.path.join(dest, f)
        if os.path.isfile(src_file):
            shutil.copyfile(src_file, dest_file)

def decode_version(s):
    def next_version(ss):
        m = re.search('(?P<major>\\d+)(?P<minor>(\\.\\d+)*)$', ss)
        if (m):
          return [m.groupdict()['major'], m.groupdict()['minor']]
        else:
          print("Could not parse version ", ss)
    version = []
    minor = s
    while(len(minor) > 0):
        major, minor = next_version(minor)
        version.append(int(major))
    print("Version = ", version)
    return version

def pandoc_version_check():
    version_check = subprocess.run(['pandoc', '--version'],
                                   stdout = subprocess.PIPE, stderr = subprocess.PIPE)
    if (version_check.returncode != 0):
        sys.stderr.writelines(['Error running pandoc:',
                               version_check.stderr.strip().decode('utf-8')])
        return None
    version_str = version_check.stdout.decode('utf-8').split('\n')[0].strip()
    version = decode_version(version_str)
    return (version[0] >= 2 and (version[0] > 2 or version[1] >= 18))

def main():
    source = sys.argv[1]
    version_ok = pandoc_version_check()
    if version_ok is None:
        sys.stderr.write("Error: Could not find pandoc. Try instlling pandoc.")
        return(1)
    if version_ok:
        print("Skipping pre-processing step")
        intermediate = source
    else:
      sys.stderr.write("Error: you need to update Pandoc to version 2.18 or higher.")
      return(1)
    bib = gen_refs(intermediate)
    gen_items(bib['references'])
    move_md_files()
    move_pdf_files()

if __name__ == '__main__':
    main()
